Fix ConflictChecker for READ-APPEND after OPTIMIZE

sezruby opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-io/delta/blob/master/CONTRIBUTING.md
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
  6. If applicable, include the corresponding issue number in the PR title and link it in the body.
-->

## Description

<!--
- Describe what this PR changes.
- Describe why we need the change.
 
If this PR resolves an issue be sure to include "Resolves #XXX" to correctly link and close the issue upon merge.
-->

Allow read with newly added files with dataChange=false and removed files with dataChange=false by concurrent transactions.
dataChange=false files are created by OPTIMIZE operation, so the content should be same.

The following functions are to check both cases.
- checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn
- checkForDeletedFilesAgainstCurrentTxnReadFiles

We can address the issue by
- fix `changedDataAddedFiles` to check dataChange flag of AddFile
- add `changedDataRemovedFiles` for RemoveFile

## How was this patch tested?
Unit tests

## Does this PR introduce _any_ user-facing changes?
No, a bug fix when Optimize transaction is committed while concurrent Reads.

sezruby committed:
Support Optimize - Append

Signed-off-by: Eunjin Song <sezruby@gmail.com>

tdas commented:
Can you elaborate on the issue? Which operation throws ConcurrentAppendException? Optimize or Append? And can give an example error and stacktrace? Also, can you explain the semantics of why this change is correct and maintain serializability?

sezruby commented:
Append operation throws ConcurrentAppendException.

```
io.delta.exceptions.ConcurrentAppendException: Files were added to partition [colC=1] by a concurrent update. Please try the operation again.
Conflicting commit: {"timestamp":1662599111087,"operation":"OPTIMIZE", ...
```

https://github.com/delta-io/delta/blob/2041c3b7138b9614529540613195d4afd39fde71/core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala#L177-L188
This change makes optimized files are not conflicting with read operation for Append operation. 

This is the repro:
```scala
// session1 - performing OPTIMIZE
(1 to 10).foreach { i => 
  println(spark.sql(s"OPTIMIZE delta.`$dataPath`").collect.toSeq)
}
```

```scala
// session 2 - performing APPEND
(1 to 10).foreach { i => 

spark.read.format("delta").load(dataPath).limit(10).write.mode("append").format("delta").partitionBy("colC").save(dataPath)
}
```

tdas commented:
@sezruby Thank you for the explanation, and the repro. So its a problem not with blind appends but read-then-append.

I have think about it, whether this change has other unintended consequence in other kind of workloads. Actually can you provide a logical argument why this change will not produce unintended consequence in other combination of operations .. like DELETE/UPDATE/MERGE + OPTIMIZE?

sezruby commented:
I found that there's still can be a conflict if concurrent transaction reads a RemoveFile from Optimize. 

io.delta.exceptions.ConcurrentDeleteReadException: This transaction attempted to read one or more files that were deleted (for example colC=0/part-00000-cf1e16ab-27b2-4c9c-b5e1-bccfb0e79a58.c000.snappy.parquet in partition [colC=0]) by a concurrent update. Please try the operation again.
Conflicting commit: {"timestamp":1663052151972,"operation":"OPTIMIZE"

https://github.com/delta-io/delta/blob/d2785aa93e8f2f5c70a3351ddc4995e2c943897b/core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala#L226-L235

I think we can add `changedDataRemovedFiles` like `changedDataAddedFiles` and use it for the check function.
We could allow Read to the previous version before OPTIMIZE without any serialization issue if the transaction doesn't update the existing data. With the fix, we could support concurrent OPTIMZIE while performing append/insert only operations which utilizes existing data for appending new rows.

For other type of operation which may have RemoveFile (DELETE UPDATE MERGE OPTIMIZE), it will fail with the following check:

https://github.com/delta-io/delta/blob/d2785aa93e8f2f5c70a3351ddc4995e2c943897b/core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala#L250-L266

sezruby committed:
Fix read-delete for optimize

Signed-off-by: Eunjin Song <sezruby@gmail.com>

sezruby committed:
Fix tests

Signed-off-by: Eunjin Song <sezruby@gmail.com>

scottsand-db commented:
Wouldn't this PR https://github.com/delta-io/delta/pull/1262 help with this?

sezruby commented:
@scottsand-db Seems the PR allows to switch isolation level to WriteSerializable. Not sure any other changes will be delivered, but with the current code, the issue still exists:
 
https://github.com/delta-io/delta/blob/d2785aa93e8f2f5c70a3351ddc4995e2c943897b/core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala#L181-L188

it still checks `changedDataAddedFiles` with WriteSerializable.

sezruby commented:
@tdas RFR

sezruby commented:
Seems the PR fixes https://github.com/delta-io/delta/issues/326

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:
```diff
@@ -96,8 +96,9 @@ private[delta] class WinningCommitSummary(val actions: Seq[Action], val commitVe
-    addedFiles
+    addedFiles.filter(_.dataChange)
   }
+  val changedDataRemovedFiles: Seq[RemoveFile] = removedFiles.filter(_.dataChange)
```
why not play it safe and perform the same `isBlindAppendOption.getOrElse(false)` as above?

sezruby replied:
I think we should detect any removeFile when isBlinedAppendOption=true as it's not blind append apparently. What do you think ?

scottsand-db replied:
That makes sense

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:
```diff
@@ -96,8 +96,9 @@ private[delta] class WinningCommitSummary(val actions: Seq[Action], val commitVe
     Seq()
   } else {
-    addedFiles
+    addedFiles.filter(_.dataChange)
```
this is worth an inline comment explaining why we want this. of course, with the variable name being `changedDataAddedFiles`, this seems obvious. but, obviously, it wasn't :) that's why you are fixing this now :) 

so if you could add an inline comment explaining why this is necessary, e.g. perhaps even with an example, that would be great

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:
```diff
@@ -96,8 +96,9 @@ private[delta] class WinningCommitSummary(val actions: Seq[Action], val commitVe
-    addedFiles
+    addedFiles.filter(_.dataChange)
   }
+  val changedDataRemovedFiles: Seq[RemoveFile] = removedFiles.filter(_.dataChange)
```
same here. inline comment.

scottsand-db commented on core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:
```diff
@@ -231,16 +232,17 @@ private[delta] class ConflictChecker(
       val readFilePaths = currentTransactionInfo.readFiles.map(
         f => f.path -> f.partitionValues).toMap
-      val deleteReadOverlap = winningCommitSummary.removedFiles
+      val deleteReadOverlap = winningCommitSummary.changedDataRemovedFiles
```
update the scaladoc for this method. we aren't just checking RemoveFile actions. we are checking RemoveFIle actions that actually change data (e.g. ignore OPTIMIZE)

scottsand-db reviewed (changes requested):
Looks great! Thanks for this change! tests look good.

Can you please update scaladoc for method `checkForDeletedFilesAgainstCurrentTxnDeletedFiles` explaining that we check for conflicts of remove files regardless of datachange status.

scottsand-db commented:
@tdas I've taken a first pass. Can you also please take a look?

prakharjain09 commented on core/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:
```diff
@@ -231,16 +232,17 @@ private[delta] class ConflictChecker(
       val readFilePaths = currentTransactionInfo.readFiles.map(
         f => f.path -> f.partitionValues).toMap
-      val deleteReadOverlap = winningCommitSummary.removedFiles
+      val deleteReadOverlap = winningCommitSummary.changedDataRemovedFiles
```
Ignoring the dataChange=false files can cause serializability issues. A txn which is trying to commit relies on the compacted files and so it indirectly relies on new files created by OPTIMIZE.

Consider the following scenario:

    
    Suppose table has 2 files initially: f1, f2.
    t0 -------------   TXN-1 Start:   	            Reads F1 and Wants to Write F11
    t1 ---             OPTIMIZE-1 Starts and Commits:   Compacts F1, F2 to F3
    t2 -------------   TXN-1 Commits:   	            Tries to commit [AddFile-F11, RemoveFile-F1] (This would have failed in master because of
                                                        ConcurrentDeleteReadException around file f1) (This commits successfully with new changes.)

At time t2 when TXN-1 tries to commit, it depends on file f1 and file f1 shouldn't change concurrently. In this case f1 was just rewritten to f3 by the concurrent OPTIMIZE and so this way we have a transitive dependency on f3 which we are missing here.

If we allow such TXN-1 to commit here, we could have issues like this:

    Suppose table has 2 files initially: f1, f2.
    t0 -------------   TXN-1 Start:   	            Reads F1 and Wants to Write F11
    t1 ---             OPTIMIZE-1 Starts and Commits:   Compacts F1, F2 to F3
    t1.5 ------        TXN-2 starts and commits:        It deletes F3. i.e. it commits RemoveFile-F3
    t2 -------------   TXN-1 Commits:   	            Tries to commit F11 and it works as it doesn't conflict with OPTIMIZE (with new logic) and it doesn't conflict with TXN-2 (TXN-2 just deleted F3 which TXN-1 was not reading directly).

Here at time t2, we will allow TXN-1 to commit successfully (after proposed changes) although TXN-1 can't be serialized after TXN-2. So ideally TXN-1 should have failed.

scottsand-db replied:
@prakharjain09 - this makes sense to me. thanks for reviewing this PR and lending a hand.

@sezruby do you understand + agree with this counter-example?

sezruby replied:
Yes I understand the case. But OPTIMIZE-APPEND is common case for many users.
If there are frequent append jobs to a partition 24/7, we cannot run optimize for the partition.
Can we add additional check for the case? 

We have readPredicates used in `checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn`  and we can detect if there is a data change commit in each partition based on it. 
I think partition level detection could be feasible.  (I was thinking tracking optimized files using BF, but checking partition would be simple/enough)

What do you think about this approach? keep the current change + partition level dataChange='true' detection

sezruby replied:
@scottsand-db @prakharjain09 I updated the PR with additional check & test for the case. Could you have a look at it? 
(I haven't addressed comment related change yet)

prakharjain09 replied:
@sezruby Could you give some example scenarios (like the one in this [comment](https://github.com/delta-io/delta/pull/1305#discussion_r986304404)) which shows the behavior before your changes and after your changes.

I think the main issue that you are trying to fix is: When Optimize commits, it affects a concurrent Append/Update/Delete/Merge query that is in progress and wants to commit due to checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn check? Is my understanding correct?

sezruby replied:
I'd like to support OPTIMIZE - APPEND scenario only. 
Other operation should fail as we cannot track modified/removed rows from optimized files and it's acceptable.
Others with RemoveFile will conflict with optimize commit in checkForDeletedFilesAgainstCurrentTxnDeletedFiles and fail.

```
Suppose table has 2 files initially: f1, f2.
t0 -------------   TXN-1 Start:   	            Reads F1 and Wants to Write F11
t1 ---             OPTIMIZE-1 Starts and Commits:   Compacts F1, F2 to F3
t1.5 ------        TXN-2 starts and commits:        It deletes F3. i.e. it commits RemoveFile-F3
t2 -------------   TXN-1 Commits:   	            Tries to commit F11, but it fails in checkForDeletedFilesAgainstCurrentTxnReadFiles, as winningCommitSummary.changedDataRemovedFiles with readPredicate filter is not empty.
```

In current implementation, 
`checkForAddedFilesThatShouldHaveBeenReadByCurrentTxn` checks newly added files using read predicates.
`checkForDeletedFilesAgainstCurrentTxnReadFiles` checks newly removed files only for readFiles of the current transaction.

I added read predicate check in `checkForDeletedFilesAgainstCurrentTxnReadFiles` so that we could detect removed files other than readFiles.

sezruby replied:
I think this (APPEND_start / OPTIMIZE - DELETE / APPEND_end) is only case that we need to check "RemoveFiles" other than readFiles. Using data filter, there could be other files not in readFiles list (but in the same partition), so the additional check would make more unnecessary(?) conflicts. (e.g. fileA only contains id=1, fileB only contains id=2 only and with a data filter id=1, fileB is not in readFiles)

Do you have any other idea how we could support OPTIMIZE-APPEND?

scottsand-db replied:
Hi @sezruby - things are getting a little complicated here :) 

Can we transition this to a more high-level discussion on the PR (not on this code) explaining the higher-level explanation/reasoning for this PR.

Inside of `checkForDeletedFilesAgainstCurrentTxnReadFiles` you seem to be checking the read predicates ... against a winningCommit's deleted files ....

@prakharjain09 is this valid?

scottsand-db replied:
@sezruby can you please clarify what you mean by APPENDs?

`Reads F1 and Wants to Write F11`

Why are you reading first before these appends? Why not do BLIND appends?

sezruby committed:
Add check

Signed-off-by: Eunjin Song <sezruby@gmail.com>

sezruby committed:
Style add

zmeir commented:
@sezruby Any news on this PR? We're running into this issue a lot, having a streaming job that writes to the same partition 24/7, and so we are having a hard time running OPTIMIZE on that partition.

Also, did you see https://github.com/delta-io/delta/pull/626 ?

sezruby commented:
> @sezruby Any news on this PR? We're running into this issue a lot, having a streaming job that writes to the same partition 24/7, and so we are having a hard time running OPTIMIZE on that partition.
> 
> Also, did you see #626 ?

Hi, you can cherry-pick this PR to support OPTIMIZE - APPEND. 
This PR contains 2 parts
- changedAddedFiles in newly added file check
  - with this, APPEND operation would succeed, as newly added files by OPTIMIZE are "dataChange=false"
- removedFiles check with readPredicates
  - master version only checks if there's newly removed files among only "readFiles" of the transaction
  - so, if there's newly deleted file after OPTIMIZE, it cannot be detected when APPEND  https://github.com/delta-io/delta/pull/1305#discussion_r986304404
  - To prevent this, I added the additional check using readPredicates, to see if there's other "newly removed files" in each read partition directory.
  - This may cause unnecessary conflicts between delete - append, I think that is very rare.
    - t0 txn0 APPEND: read file2 using data filter 
    - t1 txn1 APPEND: add file1 - commit
    - t2 txn2 DELETE: file1 - commit 
    - t3 txn0 APEND: commit => no conflict with master version, but with this PR, removed file1 can cause conflict

zmeir commented:
Thanks @sezruby !

Yeah I figured I can do that, but was hoping I could simplify my build by using an existing version artifact.

_<rolls up sleeves>_ Oh well...

foss-contributor commented:
@sezruby, thank you for your contribution! This pull request has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this pull request or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!

felipepessoto commented:
This is still valid improvement
